FirebaseStorage/Sources/StorageMetadata.swift

```diff
@@ -162,80 +105,199 @@ import FirebaseStorageInternal
+    let clone = StorageMetadata(dictionary: dictionaryRepresentation())
+    clone.initialMetadata = initialMetadata
+    return clone
   }
 
   @objc override open func isEqual(_ object: Any?) -> Bool {
     guard let ref = object as? StorageMetadata else {
       return false
     }
-    return impl.isEqual(ref.impl)
+    return ref
+      .dictionaryRepresentation() as NSDictionary == dictionaryRepresentation() as NSDictionary
   }
 
   @objc override public var hash: Int {
-    return impl.hash
+    return (dictionaryRepresentation() as NSDictionary).hash
   }
 
   @objc override public var description: String {
-    return impl.description
+    return "\(type(of: self)) \(dictionaryRepresentation())"
   }
 
   // MARK: - Internal APIs
 
-  internal let impl: FIRIMPLStorageMetadata
-
-  internal init(impl: FIRIMPLStorageMetadata) {
-    self.impl = impl
-    bucket = impl.bucket
-    md5Hash = impl.md5Hash
-    generation = impl.generation
-    metageneration = impl.metageneration
-    size = impl.size
-    timeCreated = impl.timeCreated
-    updated = impl.updated
-    if let ref = impl.storageReference {
-      storageReference = StorageReference(ref)
-    } else {
-      storageReference = nil
-    }
-    isFile = impl.isFile
-    isFolder = impl.isFolder
+  internal func updatedMetadata() -> [String: AnyHashable] {
+    return remove(matchingMetadata: dictionaryRepresentation(), oldMetadata: initialMetadata)
+  }
+
+  internal enum StorageMetadataType {
+    case unknown
+    case file
+    case folder
+  }
+
+  internal var fileType: StorageMetadataType
+
+  // MARK: - Private APIs and data
+
+  private var initialMetadata: [String: AnyHashable]
+
+  private static func intFromDictionaryValue(_ value: Any?) -> Int64 {
+    if let value = value as? Int { return Int64(value) }
+    if let value = value as? Int64 { return value }
+    if let value = value as? String { return Int64(value) ?? 0 }
+    return 0
+  }
+
+  private static var dateFormatter: DateFormatter = {
+    let dateFormatter = DateFormatter()
+    dateFormatter.locale = Locale(identifier: "en_US_POSIX") // set locale to reliable US_POSIX
+    dateFormatter.dateFormat = "yyyy'-'MM'-'dd'T'HH':'mm':'ss.SSSZZZZZ"
+    dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
+    return dateFormatter
+  }()
+
+  private static func dateFromRFC3339String(_ value: Any?) -> Date? {
+    if let stringValue = value as? String {
+      return dateFormatter.date(from: stringValue) ?? nil
+    }
+    return nil
+  }
+
+  internal static func RFC3339StringFromDate(_ date: Date) -> String {
+    return dateFormatter.string(from: date)
+  }
+
+  private func remove(matchingMetadata: [String: AnyHashable],
+                      oldMetadata: [String: AnyHashable]) -> [String: AnyHashable] {
+    var newMetadata = matchingMetadata
+    for (key, oldValue) in oldMetadata {
+      guard let newValue = newMetadata[key] else {
+        newMetadata[key] = NSNull()
```

Comment:
nit: consider leaving a comment for why we're using `NSNull` here instead of leaving it `nil`, especially since `nil` is used in other parts of this loop.

Reply:
Done